akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntitiesShardStore.scala

```diff
@@ -0,0 +1,148 @@
+import akka.persistence.DeleteMessagesSuccess
+import akka.persistence.DeleteSnapshotsFailure
+import akka.persistence.DeleteSnapshotsSuccess
+import akka.persistence.PersistentActor
+import akka.persistence.RecoveryCompleted
+import akka.persistence.SaveSnapshotFailure
+import akka.persistence.SaveSnapshotSuccess
+import akka.persistence.SnapshotOffer
+import akka.persistence.SnapshotSelectionCriteria
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object EventSourcedRememberEntitiesShardStore {
+
+  /**
+   * A case class which represents a state change for the Shard
+   */
+  sealed trait StateChange extends ClusterShardingSerializable
+
+  /**
+   * Persistent state of the Shard.
+   */
+  final case class State private[akka] (entities: Set[EntityId] = Set.empty) extends ClusterShardingSerializable
+
+  /**
+   * `State` change for starting a set of entities in this `Shard`
+   */
+  final case class EntitiesStarted(entities: Set[EntityId]) extends StateChange
+
+  case object StartedAck
+
+  /**
+   * `State` change for an entity which has terminated.
+   */
+  final case class EntitiesStopped(entities: Set[EntityId]) extends StateChange
+
+  def props(typeName: String, shardId: ShardRegion.ShardId, settings: ClusterShardingSettings): Props =
+    Props(new EventSourcedRememberEntitiesShardStore(typeName, shardId, settings))
+}
+
+/**
+ * INTERNAL API
+ *
+ * Persistent actor keeping the state for Akka Persistence backed remember entities (enabled through `state-store-mode=persistence`).
+ *
+ * @see [[ClusterSharding$ ClusterSharding extension]]
+ */
+@InternalApi
+private[akka] final class EventSourcedRememberEntitiesShardStore(
+    typeName: String,
+    shardId: ShardRegion.ShardId,
+    settings: ClusterShardingSettings)
+    extends PersistentActor
+    with ActorLogging {
+
+  import EventSourcedRememberEntitiesShardStore._
+  import settings.tuningParameters._
+
+  log.debug("Starting up EventSourcedRememberEntitiesStore")
+  private var state = State()
+  override def persistenceId = s"/sharding/${typeName}Shard/$shardId"
+  override def journalPluginId: String = settings.journalPluginId
+  override def snapshotPluginId: String = settings.snapshotPluginId
+
+  override def receiveRecover: Receive = {
+    case EntitiesStarted(ids)              => state = state.copy(state.entities.union(ids))
+    case EntitiesStopped(ids)              => state = state.copy(state.entities -- ids)
+    case SnapshotOffer(_, snapshot: State) => state = snapshot
+    case RecoveryCompleted =>
+      log.debug("Recovery completed for shard [{}] with [{}] entities", shardId, state.entities.size)
+  }
+
+  override def receiveCommand: Receive = {
+
+    case RememberEntitiesShardStore.Update(started, stopped) =>
+      val events =
+        (if (started.nonEmpty) EntitiesStarted(started) :: Nil else Nil) :::
+        (if (stopped.nonEmpty) EntitiesStopped(stopped) :: Nil else Nil)
+      var left = events.size
+      persistAll(events) { _ =>
```

Comment:
There could be a risk that the AtomicWrite of events exceeds what the journal can handle. Cassandra has a limit for that, I think. Maybe we can split this up in several `persistAll` with chunks of X events? Since this is classic PersistentActor I think it's fine to make multiple persistAll calls from same command.

Reply:
Yep, we discussed this at some point but never dealt with it, could it be an issue for the ddata store as well, with too big an update?

Reply:
> could it be an issue for the ddata store as well, with too big an update

I don't think that will be a problem. The write message can't be bigger than the full gossip message so would hit similar limits anyway.